src/xdg-autostart-generator/xdg-autostart-service.c

```diff
@@ -0,0 +1,655 @@
+                *func = xdg_config_parse_string;
+                *data = &service->gnome_autostart_phase;
+        } else if (streq(lvalue_stripped, "X-GNOME-HiddenUnderSystemd")) {
+                /* FIXME: X-systemd-skip? */
+                *func = xdg_config_parse_bool;
+                *data = &service->systemd_skip;
+        }
+        /* Parse Icon for %i? */
+
+        return 1;
+}
+
+XdgAutostartService *xdg_autostart_service_parse_desktop(const char *path) {
+        _cleanup_(xdg_autostart_service_freep) XdgAutostartService *service = NULL;
+        int r;
+
+        service = new0(XdgAutostartService, 1);
+        if (!service)
+                return NULL;
+
+        service->path = strdup(path);
+
+        r = config_parse(NULL, service->path, NULL,
+                         "Desktop Entry\0",
+                         config_item_xdg_lookup, NULL,
+                         CONFIG_PARSE_WARN, service);
+        /* If parsing failed, only hide the file so it will still mask others. */
+        if (r < 0) {
+                log_warning_errno(r, "failed to parse %s, ignoring it", service->path);
+                service->hidden = 1;
+        }
+
+        return TAKE_PTR(service);
+}
+
+int xdg_autostart_format_exec_start(
+                const char *exec,
+                char **ret_exec_start) {
+
+        _cleanup_strv_free_ char **exec_split = NULL;
+        _cleanup_free_ char *exec_start = NULL;
+        unsigned n, i;
+        bool first_arg;
+        int r;
+
+        /*
+         * Unfortunately, there is a mismatch between systemd's idea of $PATH
+         * and XDGs. i.e. we need to ensure that we have an absolute path to
+         * support cases where $PATH has been modified from the default set.
+         *
+         * Note that this is only needed for development environments though;
+         * so while it is important, this should have no effect in production
+         * environments.
+         *
+         * To be compliant with the XDG specification, we also need to strip
+         * certain parameters and such. Doing so properly makes parsing the
+         * command line unavoidable.
+         *
+         * NOTE: Technically, XDG only specifies " as quotes, while this also
+         *       accepts '.
+         */
+        exec_split = strv_split_full(exec, WHITESPACE, SPLIT_QUOTES | SPLIT_RELAX);
+        if (!exec_split)
+                return -ENOMEM;
+
+        if (strv_isempty (exec_split)) {
+                log_info("Exec line is empty");
+                return -EINVAL;
+        }
+
+        first_arg = 1;
+        for (i = n = 0; exec_split[i]; i++) {
+                _cleanup_free_ char *orig = NULL;
+                _cleanup_free_ char *c = NULL;
+                _cleanup_free_ char *raw = NULL;
+                _cleanup_free_ char *p = NULL;
+                _cleanup_free_ char *escaped = NULL;
+                _cleanup_free_ char *quoted = NULL;
+
+                orig = TAKE_PTR(exec_split[i]);
```

Comment:
It seems `TAKE_PTR` here isn't compatible with `strv_free` because it replaces `exec_split[i]` with `NULL` and `strv_free` can't get past it to fully free `exec_split`. To keep generating a seed corpora I crudely "fixed" all the memory leaks the fuzzer was able to find with
```diff
@@ -389,7 +388,7 @@ int xdg_autostart_format_exec_start(
                 _cleanup_free_ char *escaped = NULL;
                 _cleanup_free_ char *quoted = NULL;

-                orig = TAKE_PTR(exec_split[i]);
+                orig = strdup(exec_split[i]);
                 r = cunescape(orig, 0, &c);
                 if (r < 0) {
                         return log_debug_errno(r, "Failed to unescape '%s': %m", orig);
@@ -409,6 +408,7 @@ int xdg_autostart_format_exec_start(
                         if (!escaped)
                                 return log_oom();

+                        free(exec_split[n]);
                         exec_split[n++] = TAKE_PTR(escaped);
                         continue;
                 }
@@ -451,6 +451,7 @@ int xdg_autostart_format_exec_start(
                 if (!quoted)
                         return log_oom();

+                free(exec_split[n]);
                 exec_split[n++] = TAKE_PTR(quoted);
         }
```

Reply:
Oh, oops. Yeah that can actually happen in error conditions. No need for the `orig` variable then.